yarn-project/foundation/src/fields/bls12_fields.ts

```diff
@@ -0,0 +1,465 @@
+/* eslint-disable camelcase */
+import { bls12_381 } from '@noble/curves/bls12-381';
+import { inspect } from 'util';
+
+import { toBigIntBE, toBufferBE } from '../bigint-buffer/index.js';
+import { randomBytes } from '../crypto/random/index.js';
+import { hexSchemaFor } from '../schemas/utils.js';
+import { BufferReader } from '../serialize/buffer_reader.js';
+import { TypeRegistry } from '../serialize/type_registry.js';
+import { Fr } from './fields.js';
+
+/**
+ * Represents a field derived from BLS12Field.
+ */
+type BLS12DerivedField<T extends BLS12Field> = {
+  new (value: any): T;
+  /**
+   * All derived fields will specify MODULUS and SIZE_IN_BYTES.
+   */
+  MODULUS: bigint;
+  SIZE_IN_BYTES: number;
+};
+
+/**
+ * Base BLS12field class.
+ * Conversions from Buffer to BigInt and vice-versa are not cheap.
+ * We allow construction with either form and lazily convert to other as needed.
+ * We only check we are within the field modulus when initializing with bigint.
```

Comment:
This was copied over from `fields.ts` - though in testing I found no discernable slowdown when initiating a field as assigning its buffer when inputting a bigint and vice versa. So I have since changed the constructor to assign both the bigint and buffer forms.